File: app/src/main/java/net/nyx/printerclient/BaseActivity.java
package net.nyx.printerclient;

import android.app.ProgressDialog;
import android.os.Handler;
import android.os.Looper;
import androidx.appcompat.app.AppCompatActivity;

public abstract class BaseActivity extends AppCompatActivity {

    protected Handler handler = new Handler(Looper.getMainLooper());
    private ProgressDialog mDialog;

    public void showDialog() {
        if (isFinishing() || isDestroyed()) {
            return;
        }

        handler.post(() -> {
            if (isFinishing() || isDestroyed()) {
                return;
            }

            if (mDialog == null) {
                mDialog = new ProgressDialog(this);
                mDialog.setMessage("loading...");
                mDialog.setIndeterminate(false);
                mDialog.setCancelable(false);
            }
            if (!mDialog.isShowing()) {
                mDialog.show();
            }
        });
    }

    public void hideDialog() {
        if (isFinishing() || isDestroyed()) {
            return;
        }
        if (mDialog == null || !mDialog.isShowing()) {
            return;
        }
        mDialog.dismiss();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mDialog != null && mDialog.isShowing()) {
            hideDialog();
        }
        mDialog = null;
    }
}

File: app/src/main/java/net/nyx/printerclient/BaseActivity.kt
package net.nyx.printerclient

import android.os.Handler
import android.os.Looper
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import androidx.appcompat.app.AlertDialog

abstract class BaseActivity : AppCompatActivity() {

    protected val handler = Handler(Looper.getMainLooper())
    private var mDialog: AlertDialog? = null

    fun showDialog() {
        if (isFinishing || isDestroyed) {
            return
        }

        handler.post {
            if (isFinishing || isDestroyed) {
                return@post
            }

            if (mDialog == null) {
                mDialog = MaterialAlertDialogBuilder(this)
                    .setMessage("loading...")
                    .setCancelable(false)
                    .create()
            }
            mDialog?.takeIf { !it.isShowing }?.show()
        }
    }

    fun hideDialog() {
        if (isFinishing || isDestroyed) {
            return
        }
        mDialog?.takeIf { it.isShowing }?.dismiss()
    }

    override fun onDestroy() {
        super.onDestroy()
        mDialog?.takeIf { it.isShowing }?.let {
            hideDialog()
        }
        mDialog = null
    }
} 